pytracking/django.py

```diff
@@ -0,0 +1,140 @@
+from django.conf import settings
+from django.http import (
+    HttpResponseRedirect, Http404, HttpResponse)
+from django.views.generic import View
+from ipware.ip import get_ip
+from pytracking.tracking import (
+    get_configuration, TRACKING_PIXEL, PNG_MIME_TYPE)
+
+
+class TrackingView(View):
+    """Base Tracking View.
+
+    Subclasses should override notify_* methods.
+    """
+
+    def notify_tracking_event(self, tracking_result):
+        """Called once the tracking link has been decoded, and before
+        responding with a redirect or a tracking pixel.
+
+        :param tracking_result: An instance of TrackingResult.
+        """
+        pass
+
+    def notify_decoding_error(self, exception):
+        """Called when a decoding error occurs, and before
+        responding with a 404.
+
+        :param exception: The exception that was raised when trying to decode a
+            tracking link.
+        """
+        pass
+
+    def get_configuration(self):
+        """Returns a Configuration instance built from
+        settings.PYTRACKING_CONFIGURATION.
+
+        Override this method if you want to build your own Configuration
+        instance.
+        """
+        return get_configuration_from_settings()
+
+
+class ClickTrackingView(TrackingView):
+    """View that decodes a URL, call notify_tracking_event, and then sends a
+    302 redirect to the tracked URL.
+
+    This view captures the user agent and the ip of the client that makes a
+    request and puts this information in the request_data dictionary under the
+    ``user_agent`` and ``user_ip`` keys.
+
+    If a decoding error occurs, notify_decoding_error is called and a 404 is
+    returned.
+
+    If no tracking url is present in the decoded URL, a 404 is returned.
+    """
+
+    def get(self, request, path):
+        configuration = self.get_configuration()
+
+        try:
+            tracking_result = get_tracking_result(
+                request, path, False, configuration)
+        except Exception as e:
+            self.notify_decoding_error(e)
```

Comment:
The open and click views are separated for now (you need to create one view for each event) so each view will only receive a decoding error if it's an open or click decoding error. If we ever implement mixins though, it would make sense I think to have two different decoding error methods. Or am I missing something?